scripts/communications/notification_manager.py

```
@@ -0,0 +1,1382 @@
+            
+        Raises:
+            URLValidationError: If URL is invalid or not allowed
+        """
+        if not url:
+            raise URLValidationError("URL cannot be empty")
+        
+        # Parse URL
+        try:
+            parsed_url = urlparse(url)
+        except Exception as e:
+            raise URLValidationError(f"Invalid URL format: {str(e)}")
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)